File: Assets/Bundles/Client/Resources/TS/Controllers/Global/Inventory/InventoryController.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local ClientSignals = require("Client/TS/ClientSignals").ClientSignals
local CharacterEntity = require("Shared/TS/Entity/Character/CharacterEntity").CharacterEntity
local Game = require("Shared/TS/Game").Game
local Inventory = require("Shared/TS/Inventory/Inventory").Inventory
local ItemStack = require("Shared/TS/Inventory/ItemStack").ItemStack
local Network = require("Shared/TS/Network").Network
local _UserInput = require("Shared/TS/UserInput/init")
local Keyboard = _UserInput.Keyboard
local Mouse = _UserInput.Mouse
local Bin = require("Shared/TS/Util/Bin").Bin
local Signal = require("Shared/TS/Util/Signal").Signal
local InventoryController
do
	InventoryController = setmetatable({}, {
		__tostring = function()
			return "InventoryController"
		end,
	})
	InventoryController.__index = InventoryController
	function InventoryController.new(...)
		local self = setmetatable({}, InventoryController)
		return self:constructor(...) or self
	end
	function InventoryController:constructor()
		self.inventories = {}
		self.HeldSlotChanged = Signal.new()
		self.LocalInventoryAdded = Signal.new()
	end
	function InventoryController:OnStart()
		Network.ServerToClient.UpdateInventory.Client:OnServerEvent(function(dto)
			local inv = self:GetInventory(dto.id)
			if not inv then
				inv = Inventory.new(dto.id)
				local _inventories = self.inventories
				local _id = dto.id
				local _inv = inv
				_inventories[_id] = _inv
			end
			inv:ProcessDto(dto)
		end)
		Network.ServerToClient.SetInventorySlot.Client:OnServerEvent(function(invId, slot, itemStackDto, clientPredicted)
			local inv = self:GetInventory(invId)
			if not inv then
				return nil
			end
			if self.LocalInventory == inv and clientPredicted then
				return nil
			end
			local itemStack = if itemStackDto ~= nil then ItemStack:Decode(itemStackDto) else nil
			inv:SetItem(slot, itemStack)
		end)
		Network.ServerToClient.UpdateInventorySlot.Client:OnServerEvent(function(invId, slot, itemType, amount)
			local inv = self:GetInventory(invId)
			if not inv then
				return nil
			end
			local itemStack = inv:GetItem(slot)
			if itemStack == nil then
				return nil
			end
			if itemType ~= nil then
				itemStack:SetItemType(itemType)
			end
			if amount ~= nil then
				itemStack:SetAmount(amount)
			end
		end)
		Network.ServerToClient.SetHeldInventorySlot.Client:OnServerEvent(function(invId, slot, clientPredicted)
			local inv = self:GetInventory(invId)
			if not inv then
				return nil
			end
			if self.LocalInventory == inv and clientPredicted then
				return nil
			end
			inv:SetHeldSlot(slot)
		end)
		ClientSignals.EntitySpawn:ConnectWithPriority(0, function(event)
			if TS.instanceof(event.Entity, CharacterEntity) then
				local _inventories = self.inventories
				local _id = event.Entity:GetInventory().Id
				local _arg1 = event.Entity:GetInventory()
				_inventories[_id] = _arg1
				if event.Entity:IsLocalCharacter() then
					self:SetLocalInventory((event.Entity):GetInventory())
				end
			end
		end)
		local keyboard = Keyboard.new()
		local mouse = Mouse.new()
		local hotbarKeys = { 41, 42, 43, 44, 45, 46, 47, 48, 49 }
		keyboard.KeyDown:Connect(function(event)
			local _key = event.Key
			local hotbarIndex = (table.find(hotbarKeys, _key) or 0) - 1
			if hotbarIndex > -1 then
				self:SetHeldSlot(hotbarIndex)
				return nil
			end
			if event.Key == 31 then
				self:DropItemInHand()
				return nil
			end
		end)
		-- Scroll to select held item:
		mouse.Scrolled:Connect(function(delta)
			local _selectedSlot = self.LocalInventory
			if _selectedSlot ~= nil then
				_selectedSlot = _selectedSlot:GetSelectedSlot()
			end
			local selectedSlot = _selectedSlot
			if selectedSlot == nil then
				return nil
			end
			local inc = if delta < 0 then 1 else -1
			local trySlot = selectedSlot
			-- Find the next available item in the hotbar:
			for _ = 1, #hotbarKeys do
				trySlot += inc
				-- Clamp index to hotbar items:
				if inc == 1 and trySlot >= #hotbarKeys then
					trySlot = 0
				elseif inc == -1 and trySlot < 0 then
					trySlot = #hotbarKeys - 1
				end
				-- If the item at the given `trySlot` index exists, set it as the held item:
				local _itemAtSlot = self.LocalInventory
				if _itemAtSlot ~= nil then
					_itemAtSlot = _itemAtSlot:GetItem(trySlot)
				end
				local itemAtSlot = _itemAtSlot
				if itemAtSlot ~= nil then
					self:SetHeldSlot(trySlot)
					break
				end
			end
		end)
	end
	function InventoryController:SwapSlots(fromInventory, fromSlot, toInventory, toSlot, config)
		local fromItem = fromInventory:GetItem(fromSlot)
		local toItem = toInventory:GetItem(toSlot)
		local _fn = toInventory
		local _exp = toSlot
		local _object = {}
		local _left = "clientPredicted"
		local _result = config
		if _result ~= nil then
			_result = _result.noNetwork
		end
		_object[_left] = _result
		_fn:SetItem(_exp, fromItem, _object)
		local _fn_1 = fromInventory
		local _exp_1 = fromSlot
		local _object_1 = {}
		local _left_1 = "clientPredicted"
		local _result_1 = config
		if _result_1 ~= nil then
			_result_1 = _result_1.noNetwork
		end
		_object_1[_left_1] = _result_1
		_fn_1:SetItem(_exp_1, toItem, _object_1)
	end
	function InventoryController:CheckInventoryOutOfSync()
		if not self.LocalInventory then
			error("missing local inventory.")
		end
		Network.ClientToServer.Inventory.CheckOutOfSync.Client:FireServer(self.LocalInventory:Encode())
	end
	function InventoryController:DropItemInHand()
		local _heldItem = self.LocalInventory
		if _heldItem ~= nil then
			_heldItem = _heldItem:GetHeldItem()
		end
		local heldItem = _heldItem
		if heldItem then
			Network.ClientToServer.DropItemInHand.Client:FireServer(1)
		end
	end
	function InventoryController:SetLocalInventory(inventory)
		self.LocalInventory = inventory
		self.LocalInventoryAdded:Fire(inventory)
	end
	function InventoryController:ObserveLocalInventory(callback)
		local bin = Bin.new()
		local cleanup
		if Game.LocalPlayer.Character and TS.instanceof(Game.LocalPlayer.Character, CharacterEntity) then
			cleanup = callback(Game.LocalPlayer.Character:GetInventory())
		end
		bin:Add(Game.LocalPlayer:ObserveCharacter(function(entity)
			local _result = cleanup
			if _result ~= nil then
				_result()
			end
			if entity and TS.instanceof(entity, CharacterEntity) then
				cleanup = callback(entity:GetInventory())
			end
		end))
		bin:Add(function()
			local _result = cleanup
			if _result ~= nil then
				_result()
			end
		end)
		return bin
	end
	function InventoryController:ObserveLocalHeldItem(callback)
		local bin = Bin.new()
		local cleanup
		local invBin = Bin.new()
		bin:Add(self:ObserveLocalInventory(function(inv)
			invBin:Clean()
			if inv then
				invBin:Add(inv:ObserveHeldItem(function(itemStack)
					local _result = cleanup
					if _result ~= nil then
						_result()
					end
					cleanup = callback(itemStack)
				end))
			else
				cleanup = callback(nil)
			end
		end))
		return bin
	end
	function InventoryController:SetHeldSlot(slot)
		if self.LocalInventory == nil then
			return nil
		end
		self.LocalInventory:SetHeldSlot(slot)
		self.HeldSlotChanged:Fire(slot)
		Network.ClientToServer.SetHeldSlot.Client:FireServer(slot)
	end
	function InventoryController:GetInventory(id)
		local _inventories = self.inventories
		local _id = id
		return _inventories[_id]
	end
	function InventoryController:RegisterInventory(inv)
		local _inventories = self.inventories
		local _id = inv.Id
		local _inv = inv
		_inventories[_id] = _inv
	end
	function InventoryController:QuickMoveSlot(inv, slot)
		local itemStack = inv:GetItem(slot)
		if not itemStack then
			return nil
		end
		if slot < inv:GetHotbarSlotCount() then
			-- move to backpack
			local completed = false
			-- armor
			local itemMeta = itemStack:GetMeta()
			if not completed then
				if itemMeta.Armor then
					local armorSlot = inv.armorSlots[itemMeta.Armor.ArmorType]
					local existingArmor = inv:GetItem(armorSlot)
					if existingArmor == nil then
						self:SwapSlots(inv, slot, inv, armorSlot, {
							noNetwork = true,
						})
						completed = true
					end
				end
			end
			-- find slots to merge
			if not completed then
				do
					local i = inv:GetHotbarSlotCount()
					local _shouldIncrement = false
					while true do
						if _shouldIncrement then
							i += 1
						else
							_shouldIncrement = true
						end
						if not (i < inv:GetMaxSlots()) then
							break
						end
						local otherItemStack = inv:GetItem(i)
						local _result = otherItemStack
						if _result ~= nil then
							_result = _result:CanMerge(itemStack)
						end
						if _result then
							if otherItemStack:GetAmount() < otherItemStack:GetMaxStackSize() then
								local delta = math.min(itemStack:GetAmount(), otherItemStack:GetMaxStackSize() - otherItemStack:GetAmount())
								otherItemStack:SetAmount(otherItemStack:GetAmount() + delta)
								itemStack:Decrement(delta)
								if itemStack:IsDestroyed() then
									completed = true
									break
								end
							end
						end
					end
				end
			end
			if not completed then
				-- find empty slot
				do
					local i = inv:GetHotbarSlotCount()
					local _shouldIncrement = false
					while true do
						if _shouldIncrement then
							i += 1
						else
							_shouldIncrement = true
						end
						if not (i < inv:GetMaxSlots()) then
							break
						end
						if inv:GetItem(i) == nil then
							self:SwapSlots(inv, slot, inv, i, {
								noNetwork = true,
							})
							completed = true
							break
						end
					end
				end
			end
		else
			-- move to hotbar
			local completed = false
			local itemMeta = itemStack:GetMeta()
			-- armor
			if not completed then
				if itemMeta.Armor then
					local armorSlot = inv.armorSlots[itemMeta.Armor.ArmorType]
					local existingArmor = inv:GetItem(armorSlot)
					if existingArmor == nil then
						self:SwapSlots(inv, slot, inv, armorSlot, {
							noNetwork = true,
						})
						completed = true
					end
				end
			end
			-- find slots to merge
			if not completed then
				do
					local i = 0
					local _shouldIncrement = false
					while true do
						if _shouldIncrement then
							i += 1
						else
							_shouldIncrement = true
						end
						if not (i < inv:GetHotbarSlotCount()) then
							break
						end
						local otherItemStack = inv:GetItem(i)
						local _result = otherItemStack
						if _result ~= nil then
							_result = _result:CanMerge(itemStack)
						end
						if _result then
							if otherItemStack:GetAmount() < otherItemStack:GetMaxStackSize() then
								local delta = math.max(otherItemStack:GetMaxStackSize() - itemStack:GetAmount(), otherItemStack:GetMaxStackSize() - otherItemStack:GetAmount())
								otherItemStack:SetAmount(otherItemStack:GetAmount() + delta)
								itemStack:Decrement(delta)
								if itemStack:IsDestroyed() then
									completed = true
									break
								end
							end
						end
					end
				end
			end
			if not completed then
				-- find empty slot
				do
					local i = 0
					local _shouldIncrement = false
					while true do
						if _shouldIncrement then
							i += 1
						else
							_shouldIncrement = true
						end
						if not (i < inv:GetHotbarSlotCount()) then
							break
						end
						if inv:GetItem(i) == nil then
							self:SwapSlots(inv, slot, inv, i, {
								noNetwork = true,
							})
							completed = true
							break
						end
					end
				end
			end
		end
		Network.ClientToServer.Inventory.QuickMoveSlot.Client:FireServer(inv.Id, slot, inv.Id)
		-- SetTimeout(0.1, () => {
		-- this.CheckInventoryOutOfSync();
		-- });
	end
end
-- (Flamework) InventoryController metadata
Reflect.defineMetadata(InventoryController, "identifier", "Bundles/Client/Controllers/Global/Inventory/InventoryController@InventoryController")
Reflect.defineMetadata(InventoryController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(InventoryController, "$:flamework@Controller", Controller, { {} })
return {
	InventoryController = InventoryController,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/Imports/Core/Core~/src/Client/Controllers/Inventory/InventoryController.ts
import { Controller, OnStart } from "@easy-games/flamework-core";
import { CoreClientSignals } from "Client/CoreClientSignals";
import { CoreNetwork } from "Shared/CoreNetwork";
import { CharacterEntity } from "Shared/Entity/Character/CharacterEntity";
import { Game } from "Shared/Game";
import { Inventory } from "Shared/Inventory/Inventory";
import { ItemStack } from "Shared/Inventory/ItemStack";
import { Keyboard, Mouse } from "Shared/UserInput";
import { Bin } from "Shared/Util/Bin";
import { CanvasAPI } from "Shared/Util/CanvasAPI";
import { Signal, SignalPriority } from "Shared/Util/Signal";

@Controller({})
export class InventoryController implements OnStart {
	private inventories = new Map<number, Inventory>();
	public LocalInventory?: Inventory;

	public HeldSlotChanged = new Signal<number>();
	public LocalInventoryAdded = new Signal<Inventory>();

	private enabled = true;
	private disablers = new Set<number>();
	private disablerCounter = 1;

	private lastScrollTime = 0;
	private scrollCooldown = 0.1;

	constructor() {}

	OnStart(): void {
		CoreNetwork.ServerToClient.UpdateInventory.Client.OnServerEvent((dto) => {
			let inv = this.GetInventory(dto.id);
			if (!inv) {
				inv = new Inventory(dto.id);
				this.inventories.set(dto.id, inv);
			}
			inv.ProcessDto(dto);
		});
		CoreNetwork.ServerToClient.SetInventorySlot.Client.OnServerEvent(
			(invId, slot, itemStackDto, clientPredicted) => {
				const inv = this.GetInventory(invId);
				if (!inv) return;

				if (this.LocalInventory === inv && clientPredicted) return;

				const itemStack = itemStackDto !== undefined ? ItemStack.Decode(itemStackDto) : undefined;
				inv.SetItem(slot, itemStack);
			},
		);
		CoreNetwork.ServerToClient.UpdateInventorySlot.Client.OnServerEvent((invId, slot, itemType, amount) => {
			const inv = this.GetInventory(invId);
			if (!inv) return;

			const itemStack = inv.GetItem(slot);
			if (itemStack === undefined) return;

			if (itemType !== undefined) {
				itemStack.SetItemType(itemType);
			}
			if (amount !== undefined) {
				itemStack.SetAmount(amount);
			}
		});
		CoreNetwork.ServerToClient.SetHeldInventorySlot.Client.OnServerEvent((invId, slot, clientPredicted) => {
			const inv = this.GetInventory(invId);
			if (!inv) return;

			if (this.LocalInventory === inv && clientPredicted) return;

			inv.SetHeldSlot(slot);
		});
		CoreClientSignals.EntitySpawn.ConnectWithPriority(SignalPriority.HIGHEST, (event) => {
			if (event.entity instanceof CharacterEntity) {
				this.inventories.set(event.entity.GetInventory().Id, event.entity.GetInventory());
				if (event.entity.IsLocalCharacter()) {
					this.SetLocalInventory((event.entity as CharacterEntity).GetInventory());
				}
			}
		});

		const keyboard = new Keyboard();
		const mouse = new Mouse();

		const hotbarKeys = [
			KeyCode.Alpha1,
			KeyCode.Alpha2,
			KeyCode.Alpha3,
			KeyCode.Alpha4,
			KeyCode.Alpha5,
			KeyCode.Alpha6,
			KeyCode.Alpha7,
			KeyCode.Alpha8,
			KeyCode.Alpha9,
		];

		for (const hotbarIndex of $range(0, hotbarKeys.size() - 1)) {
			keyboard.OnKeyDown(hotbarKeys[hotbarIndex], (event) => {
				if (!this.enabled) return;
				this.SetHeldSlot(hotbarIndex);
			});
		}

		keyboard.OnKeyDown(KeyCode.Q, (event) => {
			if (!this.enabled) return;
			this.DropItemInHand();
		});

		// Scroll to select held item:
		mouse.Scrolled.Connect((delta) => {
			if (!this.enabled) return;
			if (math.abs(delta) > 0.1) return;

			const now = Time.time;
			if (now - this.lastScrollTime < this.scrollCooldown) {
				return;
			}
			if (CanvasAPI.IsPointerOverUI()) return;

			this.lastScrollTime = now;

			const selectedSlot = this.LocalInventory?.GetSelectedSlot();
			if (selectedSlot === undefined) return;

			const inc = delta < 0 ? 1 : -1;
			let trySlot = selectedSlot;

			// Find the next available item in the hotbar:
			for (const _ of $range(1, hotbarKeys.size())) {
				trySlot += inc;

				// Clamp index to hotbar items:
				if (inc === 1 && trySlot >= hotbarKeys.size()) {
					trySlot = 0;
				} else if (inc === -1 && trySlot < 0) {
					trySlot = hotbarKeys.size() - 1;
				}

				// If the item at the given `trySlot` index exists, set it as the held item:
				const itemAtSlot = this.LocalInventory?.GetItem(trySlot);
				if (itemAtSlot !== undefined) {
					this.SetHeldSlot(trySlot);
					break;
				}
			}
		});
	}

	public AddDisabler(): () => void {
		const id = this.disablerCounter;
		this.disablerCounter++;
		this.disablers.add(id);
		this.enabled = false;
		return () => {
			this.disablers.delete(id);
			if (this.disablers.size() === 0) {
				this.enabled = true;
			} else {
				this.enabled = false;
			}
		};
	}

	public SwapSlots(
		fromInventory: Inventory,
		fromSlot: number,
		toInventory: Inventory,
		toSlot: number,
		config?: {
			noNetwork: boolean;
		},
	): void {
		const fromItem = fromInventory.GetItem(fromSlot);
		const toItem = toInventory.GetItem(toSlot);

		toInventory.SetItem(toSlot, fromItem, {
			clientPredicted: config?.noNetwork,
		});
		fromInventory.SetItem(fromSlot, toItem, {
			clientPredicted: config?.noNetwork,
		});
	}

	public CheckInventoryOutOfSync(): void {
		if (!this.LocalInventory) {
			error("missing local inventory.");
		}

		CoreNetwork.ClientToServer.Inventory.CheckOutOfSync.Client.FireServer(this.LocalInventory.Encode());
	}

	public DropItemInHand(): void {
		const heldItem = this.LocalInventory?.GetHeldItem();
		if (heldItem) {
			CoreNetwork.ClientToServer.DropItemInHand.Client.FireServer(1);
		}
	}

	public SetLocalInventory(inventory: Inventory): void {
		this.LocalInventory = inventory;
		this.LocalInventoryAdded.Fire(inventory);
	}

	public ObserveLocalInventory(callback: (inv: Inventory) => CleanupFunc): Bin {
		const bin = new Bin();
		let cleanup: CleanupFunc;
		if (Game.LocalPlayer.Character && Game.LocalPlayer.Character instanceof CharacterEntity) {
			cleanup = callback(Game.LocalPlayer.Character.GetInventory());
		}

		bin.Add(
			Game.LocalPlayer.ObserveCharacter((entity) => {
				cleanup?.();
				if (entity && entity instanceof CharacterEntity) {
					cleanup = callback(entity.GetInventory());
				}
			}),
		);
		bin.Add(() => {
			cleanup?.();
		});
		return bin;
	}

	public ObserveLocalHeldItem(callback: (itemStack: ItemStack | undefined) => CleanupFunc): Bin {
		const bin = new Bin();

		let cleanup: CleanupFunc;

		const invBin = new Bin();
		bin.Add(
			this.ObserveLocalInventory((inv) => {
				invBin.Clean();
				if (inv) {
					invBin.Add(
						inv.ObserveHeldItem((itemStack) => {
							cleanup?.();
							cleanup = callback(itemStack);
						}),
					);
				} else {
					cleanup = callback(undefined);
				}
			}),
		);
		return bin;
	}

	public SetHeldSlot(slot: number): void {
		if (this.LocalInventory === undefined) return;

		this.LocalInventory.SetHeldSlot(slot);
		this.HeldSlotChanged.Fire(slot);
		CoreNetwork.ClientToServer.SetHeldSlot.Client.FireServer(slot);
	}

	public GetInventory(id: number): Inventory | undefined {
		return this.inventories.get(id);
	}

	public RegisterInventory(inv: Inventory): void {
		this.inventories.set(inv.Id, inv);
	}

	public QuickMoveSlot(inv: Inventory, slot: number): void {
		const itemStack = inv.GetItem(slot);
		if (!itemStack) return;

		if (slot < inv.GetHotbarSlotCount()) {
			// move to backpack

			let completed = false;

			// armor
			const itemMeta = itemStack.GetMeta();
			if (!completed) {
				if (itemMeta.armor) {
					const armorSlot = inv.armorSlots[itemMeta.armor.armorType];
					const existingArmor = inv.GetItem(armorSlot);
					if (existingArmor === undefined) {
						this.SwapSlots(inv, slot, inv, armorSlot, {
							noNetwork: true,
						});
						completed = true;
					}
				}
			}

			// find slots to merge
			if (!completed) {
				for (let i = inv.GetHotbarSlotCount(); i < inv.GetMaxSlots(); i++) {
					const otherItemStack = inv.GetItem(i);
					if (otherItemStack?.CanMerge(itemStack)) {
						if (otherItemStack.GetAmount() < otherItemStack.GetMaxStackSize()) {
							let delta = math.min(
								itemStack.GetAmount(),
								otherItemStack.GetMaxStackSize() - otherItemStack.GetAmount(),
							);
							otherItemStack.SetAmount(otherItemStack.GetAmount() + delta);
							itemStack.Decrement(delta);
							if (itemStack.IsDestroyed()) {
								completed = true;
								break;
							}
						}
					}
				}
			}

			if (!completed) {
				// find empty slot
				for (let i = inv.GetHotbarSlotCount(); i < inv.GetMaxSlots(); i++) {
					if (inv.GetItem(i) === undefined) {
						this.SwapSlots(inv, slot, inv, i, {
							noNetwork: true,
						});
						completed = true;
						break;
					}
				}
			}
		} else {
			// move to hotbar

			let completed = false;
			const itemMeta = itemStack.GetMeta();

			// armor
			if (!completed) {
				if (itemMeta.armor) {
					const armorSlot = inv.armorSlots[itemMeta.armor.armorType];
					const existingArmor = inv.GetItem(armorSlot);
					if (existingArmor === undefined) {
						this.SwapSlots(inv, slot, inv, armorSlot, {
							noNetwork: true,
						});
						completed = true;
					}
				}
			}

			// find slots to merge
			if (!completed) {
				for (let i = 0; i < inv.GetHotbarSlotCount(); i++) {
					const otherItemStack = inv.GetItem(i);
					if (otherItemStack?.CanMerge(itemStack)) {
						if (otherItemStack.GetAmount() < otherItemStack.GetMaxStackSize()) {
							let delta = math.max(
								otherItemStack.GetMaxStackSize() - itemStack.GetAmount(),
								otherItemStack.GetMaxStackSize() - otherItemStack.GetAmount(),
							);
							otherItemStack.SetAmount(otherItemStack.GetAmount() + delta);
							itemStack.Decrement(delta);
							if (itemStack.IsDestroyed()) {
								completed = true;
								break;
							}
						}
					}
				}
			}

			if (!completed) {
				// find empty slot
				for (let i = 0; i < inv.GetHotbarSlotCount(); i++) {
					if (inv.GetItem(i) === undefined) {
						this.SwapSlots(inv, slot, inv, i, {
							noNetwork: true,
						});
						completed = true;
						break;
					}
				}
			}
		}

		CoreNetwork.ClientToServer.Inventory.QuickMoveSlot.Client.FireServer(inv.Id, slot, inv.Id);

		// SetTimeout(0.1, () => {
		// 	this.CheckInventoryOutOfSync();
		// });
	}
}
